### PR TITLE
Generic index apply transform take2

### PR DESCRIPTION
**Pull request by zellerh**
Second attempt at this commit, after having to revert earlier
commits 24671b5b26 and 17da1026f6.

Commits in this PR:
- 722462e: Re-apply (cherry-pick) commits 24671b5b26 and 17da1026f6
- e4422be: Fix the failure that caused the revert, an MDP that is run only in the release build

The previous PR is https://github.com/greenplum-db/gpdb/pull/10711.
The revert PR is https://github.com/greenplum-db/gpdb/pull/10835.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by d** [comment]
the `join` regress test fails locally for me, did that happen when you ran `installcheck-good` locally?

**Comment by d** [comment]
For your convenience, here's the regress test diff I got locally:

```patch
--- /build/gpdb/src/test/regress/expected/join_optimizer.out	2020-09-16 20:01:35.039894009 +0000
+++ /build/gpdb/src/test/regress/results/join.out	2020-09-16 20:01:35.223873679 +0000
@@ -2971,22 +2971,22 @@
 where nt3.id = 1 and ss2.b3;
 QUERY PLAN
 ___________
- Hash Join
-   Hash Cond: (nt3.nt2_id = nt2.id)
-   ->  Gather Motion 3:1  (slice1; segments: 3)
+ Gather Motion 3:1  (slice1; segments: 3)
+   ->  Hash Join
+         Hash Cond: (nt3.nt2_id = nt2.id)
          ->  Index Scan using nt3_pkey on nt3
                Index Cond: (id = 1)
-   ->  Hash
-         ->  Result
-               Filter: ((nt2.b1 AND ((NOT (nt1.id IS NULL)))))
-               ->  Hash Left Join
-                     Hash Cond: (nt2.nt1_id = nt1.id)
-                     ->  Gather Motion 3:1  (slice2; segments: 3)
-                           ->  Seq Scan on nt2
-                     ->  Hash
-                           ->  Result
-                                 ->  Gather Motion 3:1  (slice3; segments: 3)
-                                       ->  Seq Scan on nt1
+         ->  Hash
+               ->  Broadcast Motion 3:3  (slice2; segments: 3)
+                     ->  Result
+                           Filter: ((nt2.b1 AND ((NOT (nt1.id IS NULL)))))
+                           ->  Nested Loop Left Join
+                                 Join Filter: true
+                                 ->  Redistribute Motion 3:3  (slice3; segments: 3)
+                                       Hash Key: nt2.nt1_id
+                                       ->  Seq Scan on nt2
+                                 ->  Index Scan using nt1_pkey on nt1
+                                       Index Cond: (id = nt2.nt1_id)
 GP_IGNORE:(17 rows)
 
 select nt3.id
@@ -4563,11 +4558,11 @@
 QUERY PLAN
 ___________
  Gather Motion 3:1  (slice1; segments: 3)
-   ->  Hash Left Join
-         Hash Cond: (parent.k = child.k)
+   ->  Nested Loop Left Join
+         Join Filter: true
          ->  Seq Scan on parent
-         ->  Hash
-               ->  Seq Scan on child
+         ->  Index Scan using child_k_key on child
+               Index Cond: (k = parent.k)
 GP_IGNORE:(7 rows)
 
 -- check for a 9.0rc1 bug: join removal breaks pseudoconstant qual handling
@@ -4661,11 +4656,11 @@
          Join Filter: true
          ->  Redistribute Motion 3:3  (slice2; segments: 3)
                Hash Key: ('***'::text)
-               ->  Hash Left Join
-                     Hash Cond: (uniquetbl.f1 = uniquetbl_1.f1)
+               ->  Nested Loop Left Join
+                     Join Filter: true
                      ->  Seq Scan on uniquetbl
-                     ->  Hash
-                           ->  Seq Scan on uniquetbl uniquetbl_1
+                     ->  Index Scan using uniquetbl_f1_key on uniquetbl uniquetbl_1
+                           Index Cond: (f1 = uniquetbl.f1)
          ->  Index Scan using uniquetbl_f1_key on uniquetbl uniquetbl_2
                Index Cond: (f1 = ('***'::text))
 GP_IGNORE:(13 rows)

======================================================================

```

**Comment by d** [comment]
I recall seeing exactly this test failure yesterday when I was looking at #10711. FWIW this is the regress test diff I got with #10230 

```patch
--- /build/gpdb/src/test/regress/expected/join_optimizer.out	2020-09-16 20:49:19.181068079 +0000
+++ /build/gpdb/src/test/regress/results/join.out	2020-09-16 20:49:19.353070112 +0000
@@ -2971,22 +2971,22 @@
 where nt3.id = 1 and ss2.b3;
 QUERY PLAN
 ___________
- Hash Join
-   Hash Cond: (nt3.nt2_id = nt2.id)
-   ->  Gather Motion 3:1  (slice1; segments: 3)
+ Gather Motion 3:1  (slice1; segments: 3)
+   ->  Hash Join
+         Hash Cond: (nt3.nt2_id = nt2.id)
          ->  Index Scan using nt3_pkey on nt3
                Index Cond: (id = 1)
-   ->  Hash
-         ->  Result
-               Filter: ((nt2.b1 AND ((NOT (nt1.id IS NULL)))))
-               ->  Hash Left Join
-                     Hash Cond: (nt2.nt1_id = nt1.id)
-                     ->  Gather Motion 3:1  (slice2; segments: 3)
-                           ->  Seq Scan on nt2
-                     ->  Hash
-                           ->  Result
-                                 ->  Gather Motion 3:1  (slice3; segments: 3)
-                                       ->  Seq Scan on nt1
+         ->  Hash
+               ->  Broadcast Motion 3:3  (slice2; segments: 3)
+                     ->  Result
+                           Filter: ((nt2.b1 AND ((NOT (nt1.id IS NULL)))))
+                           ->  Nested Loop Left Join
+                                 Join Filter: true
+                                 ->  Redistribute Motion 3:3  (slice3; segments: 3)
+                                       Hash Key: nt2.nt1_id
+                                       ->  Seq Scan on nt2
+                                 ->  Index Scan using nt1_pkey on nt1
+                                       Index Cond: (id = nt2.nt1_id)
 GP_IGNORE:(17 rows)
 
 select nt3.id
@@ -4563,11 +4558,11 @@
 QUERY PLAN
 ___________
  Gather Motion 3:1  (slice1; segments: 3)
-   ->  Hash Left Join
-         Hash Cond: (parent.k = child.k)
+   ->  Nested Loop Left Join
+         Join Filter: true
          ->  Seq Scan on parent
-         ->  Hash
-               ->  Seq Scan on child
+         ->  Index Scan using child_k_key on child
+               Index Cond: (k = parent.k)
 GP_IGNORE:(7 rows)
 
 -- check for a 9.0rc1 bug: join removal breaks pseudoconstant qual handling
@@ -4661,11 +4656,11 @@
          Join Filter: true
          ->  Redistribute Motion 3:3  (slice2; segments: 3)
                Hash Key: ('***'::text)
-               ->  Hash Left Join
-                     Hash Cond: (uniquetbl.f1 = uniquetbl_1.f1)
+               ->  Nested Loop Left Join
+                     Join Filter: true
                      ->  Seq Scan on uniquetbl
-                     ->  Hash
-                           ->  Seq Scan on uniquetbl uniquetbl_1
+                     ->  Index Scan using uniquetbl_f1_key on uniquetbl uniquetbl_1
+                           Index Cond: (f1 = uniquetbl.f1)
          ->  Index Scan using uniquetbl_f1_key on uniquetbl uniquetbl_2
                Index Cond: (f1 = ('***'::text))
 GP_IGNORE:(13 rows)

======================================================================

```

**Comment by d** [comment]
Hi @dgkimura and @zellerh , what's the best time I can push a (rebased and) reformatted history to your PR so that it doesn't disrupt what you're working on?

**Comment by zellerh** [comment]
> Hi @dgkimura and @zellerh , what's the best time I can push a (rebased and) reformatted history to your PR so that it doesn't disrupt what you're working on?

Anytime is good. Thanks!

**Comment by zellerh** [comment]
@d, let me know if you would prefer that I rebase the branch to after the PG merge first.

**Comment by d** [comment]
> @d, let me know if you would prefer that I rebase the branch to after the PG merge first.

Ah no worries I'll do both in one go

**Comment by d** [comment]
Hi @zellerh and @dgkimura I've pushed a (rebased and) reformatted version of this PR's history. For posterity I'm also hosting the two versions on my fork:

before the reformat: https://github.com/d/gpdb/tree/10843-backup
after: https://github.com/d/gpdb/tree/10843-fmt

Also note: I've resolved the rebase conflict by taking the Postgres 12 merge's deletion of `join_optimizer.out`. Resolving that FIXME might be a blocker for you two to continue working on this

**Comment by zellerh** [comment]
Thanks, @d! I was planning to resolve the conflict the same way. We will take care of any test failures.

File: src/test/regress/expected/gporca_optimizer.out [hunk 1]
```diff
@@ -13419,15 +13490,48 @@ from foo l1 where b in (select ab
...
-                        from foo l2 join (select *, l1.a+tbtree.b as ab from tbtree) tbtree_derived
-                                    on l2.a=tbtree_derived.a and l2.b=tbtree_derived.b
-                        where l2.c = 1
-                       );
- a | b | c 
----+---+---
-(0 rows)
+-- 16 group by columns are not a superset of the distribution columns - no index scan
+explain (costs off)
+select * from foo join (select b, count(*) as cnt from tbtree group by b) grby on foo.a=grby.cnt;
```
Comment (chrishajas) [inline src/test/regress/expected/gporca_optimizer.out:13493]: Should we also run these queries or is explain only ok?

**Review by chrishajas — APPROVED** [review]
LGTM. Out of curiosity, how did you discover the groupby dist col issue?

File: src/test/regress/expected/gporca_optimizer.out [hunk 2]
```diff
@@ -13342,16 +13342,14 @@ select * from foo join (select a, count(*) + 5 as cnt from tbitmap where tbitmap
                Filter: (a < 5)
          ->  Result
                Filter: ((count()) < 2)
-               ->  GroupAggregate
```
[the remark this answers is not in the document]
Comment (zellerh) [reply src/test/regress/expected/gporca_optimizer.out]: I'm looking into this. I now have two subsequent pipeline runs with the same code that produce different plans.

File: src/test/regress/expected/gporca_optimizer.out (hunk 1 above)
Comment (zellerh) [reply src/test/regress/expected/gporca_optimizer.out:13493]: These queries don't use the transform, so I just wanted to make sure we don't get the index nested join.

**Comment by zellerh** [comment]
> Out of curiosity, how did you discover the groupby dist col issue?

Thanks for taking another look, @chrishajas. There were two test failures, olap_group and qp_with_clause, that lead me to this problem. I'm sorry for finding these test failures late, I should have noticed them much earlier.

Hope to merge this PR once I figure out a solution to the flaky plan issue in the gporca test noted above.

**Comment by zellerh** [comment]
Good idea, @chrishajas, to add an analyze. Without the analyze, one of the alternatives ends up being pruned away. With the added analyze, we cost both alternatives and the costs are within .00004 % of each other, 432.866735 vs. 432.866888. If we continue to have this kind of problem we may have to investigate more.

**Pull request merged**